### ISSUE TITLE
Tagging restrictions preventing deployment of Firewall Policy in the landing zone

### ISSUE THREAD

**Issue by simondtaylor** [Azure/terraform-azurerm-caf-enterprise-scale#531]
### Versions

**terraform**: 1.3.5

**azure provider**: 3.22.0

**module**: caf-enterprise-scale

### Description

I have an issue deploying Azure Firewall policies due to the inability to set tags on these resources. We have implemented a policy at the root management group level to prevent deployment of non tagged resources. When deploying an Azure Firewall after the initial landing zone deployment I get an error complaining that the resource cannot be deployed due to a policy violation (no tags).

The policies applied at the root MG are shown below

```

{
    "extend_es_root": {
      "policy_assignments": ["Enforce-TLS-SSL", "Enforce-Mandatory-Tags", "Deny-RDP-SSH", "Deploy-SQL-Security", "ANS-ServiceProvider"],
      "policy_definitions": ["Enforce-RG-Tags", "Enforce-Resource-Tags", "Deny-NIC-NSG", "Deny-SSH-From-Internet", "Allowed-Tenant-Management"],
      "policy_set_definitions": ["Deny-RDP-SSH", "Enforce-Mandatory-Tags", "ANS-ServiceProvider", "NIC-Restrictions"],
      "role_definitions": [],
      "archetype_config": {
        "parameters": {},
        "access_control": {}
      }
    }
  }
```

I have tried excluding the policy using an exclusion: '/lib/archetype_exclusion_es_connectivity.tmpl.json' under the connectivity management group as per below but it **does not** take effect.

This approach would be a sledgehammer to crack a nut anyway but the best option I have so far.

```
{
    "exclude_es_connectivity": {
      "policy_assignments": ["Enforce-Mandatory-Tags"],
      "policy_definitions": [],
      "policy_set_definitions": [],
      "role_definitions": [],
      "archetype_config": {
        "parameters": {},
        "access_control": {}
      }
    }
}
```

It would be quite useful if you could exclude individual resources from tagging requirements.

#### Steps to Reproduce

1. Deploy Hub & Spoke Landing Zone with policy to prevent resources created without tags. (Azure Firewall set to false)
2. Update Azure Firewall  to 'true' and apply the changes. 
3. Deployment fails due to tagging restrictions. Screen shot below.

#### Screenshots

![Screenshot 2022-11-30 at 11 50 49](https://user-images.githubusercontent.com/32516551/204789134-b50f6d2b-a571-4895-9cd0-0fa4f53620a2.png)

**Comment by krowlandson** [comment]
Thank you for your question @simondtaylor 

Unfortunately, the approach you suggest above isn't going to help you as the archetype model is only about manipulating archetype definitions, and not creating exclusions at lower scopes.

To exclude an individual resource from an inherited Policy Assignment you would need to either [create an exclusion within the assignment](https://learn.microsoft.com/en-us/azure/governance/policy/tutorials/create-and-manage#remove-a-non-compliant-or-denied-resource-from-the-scope-with-an-exclusion) or apply a [Policy Exemption](https://learn.microsoft.com/en-us/azure/governance/policy/concepts/exemption-structure) (recommended).

You could create an exclusion within your Policy Assignment, but our recommended approach would be to create an exemption. These are not in scope of this module, so you would need to create the required exemptions by using one of the following (as deemed appropriate based on required scope):

- [azurerm_resource_policy_exemption](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/resource_policy_exemption)
- [azurerm_resource_group_policy_exemption](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/resource_group_policy_exemption)
- [azurerm_subscription_policy_exemption](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/subscription_policy_exemption)
- [azurerm_management_group_policy_exemption](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/management_group_policy_exemption)

Azure Firewall Policies support tags, with the following limitations noted in the Microsoft docs:

> Azure Firewall Policies don't support PATCH operations, which means they don't support updating tags through the portal. Instead, use the update commands for those resources.
_src: https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/tag-support#network-limitations_

Within the module, we also support adding tags to Azure Firewall Policies

https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/blob/353aaebd7585eae4cf1fc91578fb609fda653f3f/resources.connectivity.tf#L254

By default, these tags will be populated by a merge [^1] of the following input variables:

- [default_tags](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/wiki/%5BVariables%5D-default_tags)
- [configure_connectivity_resources.settings.tags](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/wiki/%5BVariables%5D-configure_connectivity_resources#settingstags)

Alternatively, you can override the tags on a specific Azure Firewall Policy using the `advanced` block in `configure_connectivity_settings`:

```hcl
    advanced = {
      custom_settings_by_resource_type = {
        azurerm_firewall_policy = {
          connectivity = {
            ukwest = { # Replace 'ukwest' with the location you are deploying this resource to
              tags= {
                Application = ""
                Criticality = ""
                Environment = ""
                Owner = ""
              }
            }
          }
        }
      }
    }
```

Based on your `Steps to Reproduce`, it sounds like this is expected behaviour. Have you tried specifying tags using one of the methods above?

If you are applying these tags to your resource but still getting blocked by the policy, we might need further information from you to determine whether the module is at fault, or the policy is incorrectly configured.

[^1]: Prior to release `v3.0.0` these values were applied in order of precedence, rather than merged.

**Comment by simondtaylor** [comment]
Thanks for the response Kevin and the recommended approach. I think the tagging overrides might work best in this situation.

**Comment by krowlandson** [comment]
Just to warn, you might get past the Azure Firewall Rules, but I assume the other resources will then also need tags to be applied?

Would you be better going for the [configure_connectivity_resources.settings.tags](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/wiki/%5BVariables%5D-configure_connectivity_resources#settingstags) option?

**Comment by simondtaylor** [comment]
Hi Kevin! Thank you for the response (again). Taking on board what you have said, I agree - it's better to use settings.tags. Just tested and its working great. Appreciate you taking the time out to respond!

Many Thanks.

**Comment by krowlandson** [comment]
Awesome, thank you for confirming @simondtaylor ... I will close the issue, but please re-open if you have any further questions regarding this matter 👍🏻
